### DISCUSSION TITLE
Automatically relate a product when it is in advanced field

### DISCUSSION THREAD

**Discussion by kseibi** [YetiForceCompany/YetiForceCRM#14911] [How to]
In YETIFORCE 6.1 we have the function to do many-to-many relation between product and sales orders, it's working fine but it needs to be added manually, is there a way to add the relationship between product and sales order right when the new sales order is created (With advanced fields)? just like the storage module, we make a goods receive note through advanced fields and it's automatically linked to the product itself, I can program it by myself but I would need some directions where to start, thank you very much.

**Comment by vovpff** [comment]
You need to write custom relation type. Yuo can start from lerning how the many-to-many relation works.

**Comment by kseibi** [comment]
I was finally able to achieve it with by adding below code to save.php, I had to change the page:
modules/Vtiger/actions/save.php and add the below code at the line 95 right below /** * {@inheritdoc} */:
```
public function saveRecord(App\Request $request)
	{
		$this->getRecordModelFromRequest($request);
		$eventHandler = $this->record->getEventHandler();
		foreach ($eventHandler->getHandlers(\App\EventHandler::EDIT_VIEW_PRE_SAVE) as $handler) {
			if (!(($response = $eventHandler->triggerHandler($handler))['result'] ?? null)) {
				throw new \App\Exceptions\NoPermittedToRecord($response['message'], 406);
			}
		}
		$this->record->save();
		$inventory = $request->get('inventory');
		$parentModuleName = $request->get('module');
		$parentModuleModel = Vtiger_Module_Model::getInstance($parentModuleName);
		$parentRecordId = (empty($request->get('record')))?$this->record->getId():$request->get('record');
        $relatedModule = "Products";
		$relatedModuleModel = Vtiger_Module_Model::getInstance($relatedModule);
		$relationModel = Vtiger_Relation_Model::getInstance($parentModuleModel, $relatedModuleModel); 

				//Replace SSingleOrders with any module name you need;
		if($parentModuleName == "SSingleOrders" && count($inventory) > 0){
            			
			for ($x = 1; $x <= count($inventory); $x++) {
				$relatedRecordId = $inventory[$x]['name'];
				$relationModel->addRelation($parentRecordId, $relatedRecordId);
			}
		}
		//Replace YOUR-MODULE-NAME with any module name you need, keep copying below code for as many modules needed;
			if($parentModuleName == "YOUR-MODULE-NAME" && count($inventory) > 0){
            			
			for ($x = 1; $x <= count($inventory); $x++) {
				$relatedRecordId = $inventory[$x]['name'];
				$relationModel->addRelation($parentRecordId, $relatedRecordId);
			}
		}
				
		if ($request->getBoolean('relationOperation')) {
			$relationId = $request->isEmpty('relationId') ? false : $request->getInteger('relationId');
			if ($relationModel = Vtiger_Relation_Model::getInstance(Vtiger_Module_Model::getInstance($request->getByType('sourceModule', 2)), $this->record->getModule(), $relationId)) {
				$relationModel->addRelation($request->getInteger('sourceRecord'), $this->record->getId());
			}
		}
	}
```

Above code will only work **if previously created a "Relation" between the module and the advanced field name module**, for example in sales order the item-name module is products, must create a relation from Settings -> Standard modules -> Modules - Relations for sales order with products module, it will an give error if the module is not previously related.

**Reply by mariuszkrzaczkowski** [reply]
Less secure code, better to do it with additional handler on "entityAfterSave"
example: https://github.com/YetiForceCompany/YetiForceCRM/blob/developer/modules/Vtiger/handlers/RecordStatusHistory.php#L41
